jupyter_ai_tools/utils.py

```diff
@@ -1,20 +1,72 @@
 import functools
 import inspect
+import os
 import typing
+from pathlib import Path
 from typing import Optional
+from urllib.parse import unquote
 
 from jupyter_server.serverapp import ServerApp
 from jupyter_server.auth.identity import User
 from pycrdt import Awareness
 
 
-async def get_serverapp():
+def get_serverapp():
     """Returns the server app from the request context"""
 
     server = ServerApp.instance()
     return server
 
 
+def normalize_filepath(file_path: str) -> str:
```

Comment:
Might benefit from caching, since this is called in numerous places.

```suggestion
@functools.lru_cache(maxsize=256)
def normalize_filepath(file_path: str) -> str:
```